### DISCUSSION TITLE
LocalStorage when working with two useStates in NextJS

### DISCUSSION THREAD

**Discussion by StefanieLoeser** [neuefische/questions#176] [Web] [answered]
Hi everybody,

I wanted to implement the localStorage functionality. I transferred the hook provided by @ThomasKirst and thought I adapted it according to the structure of my code – I have a first useState that has to stay as it contains existing data that gets rendered. The selection that can be made and gets saved in a second useState should now be stored in localStorage. Unfortunately I get a syntax error. 

Thank you for your help!

[Link to my PR](https://github.com/StefanieLoeser/capstone-project/pull/13)

This is how the code looks like:
![Bildschirmfoto 2022-09-28 um 09 28 37](https://user-images.githubusercontent.com/104068626/192715879-52180e5c-8f1e-473f-91e3-c0fbd576f1dc.png)

The error message and localStorage in DevTools:
![Bildschirmfoto 2022-09-28 um 09 49 31](https://user-images.githubusercontent.com/104068626/192720569-72e8dc18-b16d-4ba5-ba49-ea7902402ddf.png)

**Comment by ThomasKirst** [comment, marked as answer]
Instead of writing `undefined` into the localStorage, which would be the "return" of the anonymous function that you pass into `setSelectionState()` …
```js
setSelectionState(() =>
      updatedCollection.filter((record) => record.isChecked)
);
```
… pass the direct return of the filter method into your setter function:
```
setSelectionState(updatedCollection.filter((record) => record.isChecked));
```

**Reply by StefanieLoeser** [reply]
Thank you, @ThomasKirst. It worked for the state of the selection. I then realized that I do actually also need a localStorage to save which records are selection in the collection so that the checkbox stays ticked. 
I tried to refactor the code to first connect the useState with the localstorage and then update it with the fetched data but get the following error now. 

The refactored code: 
![Bildschirmfoto 2022-09-28 um 10 37 04](https://user-images.githubusercontent.com/104068626/192731383-c9da1767-39ba-4fb4-ba1a-a368b63050e9.png)

The error: 
![Bildschirmfoto 2022-09-28 um 10 34 40](https://user-images.githubusercontent.com/104068626/192731488-4468c2ae-b443-4730-9788-443ccf886412.png)

**Reply by NikRoe** [reply]
Hey Stefanie,

it looks like you made a minor change to the `useLocalStorage` hook and destructured the second parameter. The hook should start with 
```js
export default function useLocalStorage(key, selection) {
...
}
```

In your `_app.js` you can use the array `myCollection` as the initial state of your `collectionState`. Like this:
```js
const [collectionState, setCollectionState] = useLocalStorage(
    '_collection',
    myCollection
  );
```

On another note, you can delete the stories folder :)

**Reply by StefanieLoeser** [reply]
Thank you for your feedback @NikRoe. I made the changes as suggested. Now the app works again and I don't have an error but unfortunately the value change of "isChecked" doesn't get stored in the localStorage and hence doesn't get rendered when leaving and returning to the collection page. 
Do you have another idea, what might be the issue here?

**Reply by StefanieLoeser** [reply]
I found the solution myself – I had to add the "checked"-attribute to the checkbox so it actually received the info whether it is checked or not.

**Reply by NikRoe** [reply]
perfect 👍
